### PR TITLE
Imdb

### PR DESCRIPTION
**Pull request by tomkoren21**
Add IMDB metadata, model flag for choosing.

### CODE REVIEW

**Review by nir-bd — COMMENTED** [review]
This part is not clear in terms of variable naming. 
The division to train/validation could be more clear.
Please consider refactoring.
In addition, consider adding a comment for the `for...` section where building all the inputs.

**Review by doronHarnoy — APPROVED** [review]
Really good work! I left a couple of minor suggestions and questions
Generally it would be great to reduce the amount of lines in the dataset script here but i know it is tough, lets see it together maybe we can simplify some code blocks

File: examples/imdb/imdb/README.md [hunk 1]
```diff
@@ -1,3 +1,101 @@
...
+Our new model aims to fix the missing context by replacing the dense layer with 3 convulutional blocks,
+which should help in capturing a wider context.
+
+![Embedding space](images/img_11.png)
+
+We first train the model for 3 epochs to get an accuracy of 0.89 and a loss of 0.28 
+
+![Embedding space](images/img_12.png)
+
+Following the same process, we can examine the same sample, which result in a much lower loss now:
```
Comment (nir-bd) [inline examples/imdb/imdb/README.md:61]: Consider adding a comparison between the sample's loss.

File: README.md [hunk 2]
```diff
@@ -9,7 +9,9 @@ This repository provide examples to Tensorleap's integration with multiple model
...
 ## Examples 
 
-| Example                  | Task         |                  Description                                                               | 
-|--------------------------|--------------|------------------------------------------------------------------------------------------  |
-| [COCO](/examples/coco)   | Segmentation | We've used the `COCO 14` dataset with a backbone of`MobileNetV2` and a `pix2pix` upsampler |
+| Example                  | Task               |                  Description                                                               | 
+|--------------------------|--------------------|------------------------------------------------------------------------------------------  |
+| [COCO](/examples/coco)   | Segmentation       | We've used the `COCO 14` dataset with a backbone of`MobileNetV2` and a `pix2pix` upsampler |
+|--------------------------|--------------------|------------------------------------------------------------------------------------------  |
+| [IMDB](/examples/imdb)   | Sentiment Analysis | We've used a `CNN`  and a `DNN` with an embedding layer to classifiy sentiment in reviews  |
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply README.md:16]: Fixed

File: examples/imdb/imdb/.tensorleap/model.py [hunk 3]
```diff
@@ -1,10 +1,18 @@
...
     # Save it to the path supplied as an arugment (has a .h5 suffix)
-    model = tensorleap_model()
+    chosen_model = MODEL_TYPES[0]
+
+    if chosen_model == MODEL_TYPES[0]:
+        model = tensorleap_dense_model()
+    elif chosen_model == MODEL_TYPES[1]:
+        model = tensorleap_masked_conv_model()
+    elif chosen_model == MODEL_TYPES[2]:
+        model = tensorleap_conv_model()
```
[the remark this answers is not in the document]
Comment (doronHarnoy) [reply examples/imdb/imdb/.tensorleap/model.py:16]: I agree, lets discuss it all of us together

File: examples/imdb/imdb/.tensorleap/dataset.py [hunk 4]
```diff
@@ -104,18 +104,26 @@ def subset_func() -> List[SubsetResponse]:
...
-                                                       'metrics': train_metrics})
-    val = SubsetResponse(length=2*half_v_size, data={'paths': val_paths, 'gt': val_gt, 'tokenizer': tokenizer,
-                                                     'metrics': val_metrics})
+    input_keys_pre = ['', 'metrics', 'length', 'oov', 'polarity', 'subjectivity']
+    output_keys = ['paths', 'metrics', 'length', 'oov_count', 'polarity', 'subjectivity']
+    gt = ['pos', 'neg']
+    data_dict = {'train': {}, 'val': {}}
+    for in_key_pre, out_key in zip(input_keys_pre, output_keys):
+        if in_key_pre != '':
+            in_key = [gt[j] + "_" + in_key_pre for j in range(len(gt))]
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/.tensorleap/dataset.py:113]: I've added a refactoring of lines 100-110 and comments. line 113 access an auxiliary, preprocessed dictionary with keys "[POS/NEG]_METRICNAME" and arranges this in a Tensorleap format for later processing.

File: examples/imdb/imdb/.tensorleap/dataset.py [hunk 5]
```diff
@@ -61,11 +61,11 @@ def load_tokanizer(tokanizer_path: str) -> TokenizerType:
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/.tensorleap/dataset.py:61]: Done

File: examples/imdb/imdb/.tensorleap/config.toml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 projectName = "imdb"
 organization = "tom"
 framework = "tensorflow"
-datasetName = "imdb_metadata"
+datasetName = "imdb_v4"
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/.tensorleap/config.toml:4]: Done

File: examples/imdb/imdb/README.md [hunk 7]
```diff
@@ -1,3 +1,101 @@
...
+`positive` prediction made by the model. 
+
+Since the dense model makes its prediction using a single word context, it considers single words, such as
+**entertaining** or **masterpiece** as an indication for a **positive** sentiment. However, it ignores their context,
+and therefore misses that its only **mildly entertaining** or a **masterpiece because of sympathy.**
+To improve upon this we'll evaluate a second, convulutional model.
+
+### CNN
+
+Our new model aims to fix the missing context by replacing the dense layer with 3 convulutional blocks,
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/README.md:52]: Fixed

File: examples/imdb/imdb/README.md [hunk 8]
```diff
@@ -1,3 +1,101 @@
...
+![img_1.png](images/img.png)
+
+`length` - the shorter the review, the higher the loss.  
+`out-of-vocabulary` - the more out-of-vocabulary words a review has, the higher its loss  
+`subjectiveness` - the more subjective the review is the lower the loss.  
+`score confindence` - the more confident a review is (i.e. highly negative or positive) the lower the loss.  
+`polarity` - We use an external (TextBlob) polarity analysis, and show that sentences that have a neutral polarity, have a higher loss.  
+
+
+Finally, we can see that the `Tensorleap's` system unsupervised clustering is able to group together meaningful
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/README.md:85]: Done

File: examples/imdb/imdb/README.md [hunk 9]
```diff
@@ -1,3 +1,101 @@
...
+In fact, the bi-gram "entertaining mildly" is now an indicator of a negative sentiment:
+![img.png](images/img_18.png)
+
+
+#### Data exploration
+
+Using Tensorleap's dashboard we can see how our data is distributed across various features.
+Here, we've selected 5 informative features and plotted their hitogram vs. the loss:
+
+![img_1.png](images/img.png)
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/README.md:76]: Done

File: examples/imdb/imdb/.tensorleap/dataset.py [hunk 10]
```diff
@@ -144,12 +152,37 @@ def func(idx: int, subset: SubsetResponse) -> float:
...
+
+def oov_metadata(idx: int, subset: SubsetResponse) -> int:
+    return subset.data['oov_count'][idx]
+
+
+def score_metadata(idx, subset: SubsetResponse) -> int:
+    return int(subset.data['paths'][idx].split("_")[1].split(".")[0])
+
+
+def score_confidence(idx, subset: SubsetResponse) -> int:
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/.tensorleap/dataset.py:171]: Done

File: examples/imdb/imdb/README.md [hunk 11]
```diff
@@ -1,3 +1,101 @@
...
+
+We train our model for 5 epochs and log the `CategoricalCrossEntropy` loss and the `Accuracy` metric:
+
+![Training metrics](images/img_3.png)
+
+Our final accuracy is 0.89 and our loss is 0.27.
+
+#### Error analysis 
+
+To explorer the model's performance we visualize the latent space represntation using  `Tensorleap's Analyzer`,
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/README.md:32]: Done

**Comment by tomkoren21** [comment]
@nir-bd  i'm unable to answer your first comment directly but I've refactored that section and added comments

File: examples/imdb/imdb/.tensorleap/dataset.py [hunk 12]
```diff
@@ -144,12 +152,37 @@ def func(idx: int, subset: SubsetResponse) -> float:
     return func
 
 
-def gt_metadata(idx: int, subset: SubsetResponse) -> List[float]:
+def gt_sentiment(idx: int, subset: SubsetResponse) -> List[float]:
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/.tensorleap/dataset.py:155]: Added the sections with comments

File: examples/imdb/imdb/README.md (hunk 1 above)
Comment (tomkoren21) [reply examples/imdb/imdb/README.md:61]: Added the loss reduction (-67%)

File: examples/imdb/imdb/.tensorleap/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomkoren21) [reply examples/imdb/imdb/.tensorleap/model.py:16]: Agreed to switch to dictionary instead

**Pull request merged**